crates/ecs/src/systems/mod.rs

```diff
@@ -420,14 +504,77 @@ macro_rules! impl_system_parameter_function {
                         .flatten()
                         .collect()
                 }
             }
 
+            impl<'a, $([<P$parameter>]: SystemParameter,)*> IntoIterator for Query<'a, ($([<P$parameter>],)*)> {
+                type Item = ($([<P$parameter>],)*);
+                type IntoIter = QueryIterator<'a, ($([<P$parameter>],)*)>;
+
+                fn into_iter(self) -> Self::IntoIter {
+                    let base_signature: Vec<TypeId> = [$([<P$parameter>]::base_signature(),)*]
+                        .into_iter()
+                        .flatten()
+                        .collect();
+
+                    let universe = self.world.get_archetype_indices(&base_signature);
+
+                    let archetypes_indices: Vec<_> = intersection_of_multiple_sets(&[
+                        universe.clone(),
+                        $([<P$parameter>]::filter(&universe, self.world),)*
+                    ])
+                    .into_iter()
+                    .collect();
+
+                    let borrowed = (
+                        $([<P$parameter>]::borrow(self.world, &archetypes_indices).unwrap(),)*
+                    );
+
+                    let iterate_over_entities = $([<P$parameter>]::iterates_over_entities() ||)* false;
+
+                    Self::IntoIter {
+                        borrowed,
+                        iterate_over_entities,
+                        iterated_once: false,
+                    }
+                }
+            }
+
+            impl<'a, $([<P$parameter>]: SystemParameter,)*> Iterator for QueryIterator<'a, ($([<P$parameter>],)*)> {
+                type Item = ($([<P$parameter>],)*);
+
+                fn next(&mut self) -> Option<Self::Item> {
+                    // SAFETY: This is safe because the result from fetch_parameter will not outlive borrowed
+                    unsafe {
+                        if self.iterate_over_entities {
+                            while let ($(Some([<parameter_$parameter>]),)*) = (
+                                $([<P$parameter>]::fetch_parameter(&mut self.borrowed.$parameter),)*
+                            ) {
+                                if let ($(Some([<parameter_$parameter>]),)*) = (
+                                    $([<parameter_$parameter>],)*
+                                ) {
+                                    return Some(($([<parameter_$parameter>],)*));
+                                }
+                            }
+                        } else if let (false, $(Some(Some([<parameter_$parameter>])),)*) = (
+                            self.iterated_once,
+                            $([<P$parameter>]::fetch_parameter(&mut self.borrowed.$parameter),)*
+                        ) {
+                            self.iterated_once = true;
+                            return Some(($([<parameter_$parameter>],)*));
+                        }
+                        None
+                    }
```

Comment:
I thought about this too. One thing that needs to change if we do that is that we can no longer return a `SystemResult`. Right now, this error only occurs when an invalid archetype index is used.

Reply:
Hm, I was looking to see if there's something akin to an `TryIntoIterator` but I can't find anything. Since the error could only occur during borrowing, it should only be able to happen during iterator construction, right? Would it maybe be possible to change the `type IntoIter` into a result-type?

Reply:
> Since the error could only occur during borrowing, it should only be able to happen during iterator construction, right?

Yes, this can only happen during iterator construction.

> Would it maybe be possible to change the `type IntoIter` into a result-type?

No, `type IntoIter` needs to implement `Iterator<Item=Self::Output>`. We could change the `Self::Output` to a `SystemResult`, but would make `Query` very enjoying to use for the end-user.

Reply:
We could maybe do that, but then we need to dublicate the code in `into_iter` to the associated constructor-function.  Isn't the whole point of using `Query` internally to avoid duplicated code?

Reply:
You're right. I can try to implement this tomorrow.